teps/0115-tekton-catalog-git-based-versioning.md

```diff
@@ -0,0 +1,334 @@
+    - name: name
+      value: buildpacks
+    - name: version
+      value: 0.5
+```
+
+#### HTTP Endpoint
+
+In [TEP-0110][tep-0110], we added an HTTP endpoint that serves Tekton resources that users can apply directly to their
+clusters. These endpoints should continue working as is:
+
+```shell
+kubectl apply -f https://hub.tekton.dev/tekton/task/buildpacks/0.5/raw
+```
+
+### Releases
+
+Releases of resources will be done through creating tags in the Catalogs. The frequency will be monthly to match other
+Tekton projects. The tags should continue to use the simple version of [semver][semver]: `{major}.{minor}`. These tags
+will in catalogs will map to versions in the Hub e.g. tag "v0.7" in a Catalog == version "0.7" in the Hub.
+
+Note that version tags in a Catalog apply to the whole repository - this may be confusing users when there are no 
+changes in a specific resource. Regardless this is the preferred approach compared to the [alternatives][alternatives].
+
+### Migration
+
+The Hub should support both the old and new contracts to provide backwards compatibility.
+
+1. The existing Tekton Catalog - `tektoncd/catalog` - should remain as it is using the old contract. 
+2. A new Tekton Community Catalog - `tektoncd/catalog-community` - will be added that uses the new contract. This will
+be initialized with the latest versions of resources from the existing Tekton Catalog for continuity.
+3. The Hub will serve the two Catalogs - `tektoncd/catalog-community` and `tektoncd/catalog-official` - as we 
+migrate to the new contract over 9 months. During this time `tektoncd/catalog` will be read-only.
+4. After 9 months, `tektoncd/catalog` will be removed and the Hub will serve `tektoncd/catalog-community` and 
+`tektoncd/catalog-official` only.
+
+The renaming from `tektoncd/catalog` to `tekton/catalog-community` also clarifies the naming of the two Catalogs that
+Tekton will provide for official and community support tiers, as proposed in [TEP-0079][tep-0079].
+
+When [TEP-0079][tep-0079] is implementable, the Tekton Official Catalog - `tektoncd/catalog-official` - will be added 
+using the new contract.
+
+## Alternatives
+
+### Releases - Submodules in Catalogs
+
+We could require that each resource is defined in its own repository, then they are included in the Catalog as 
+[submodules][submodules]. This will ensure that each resource will have its own tag, hence its own versioning.
+However, this will be a drastic change to the Catalog organization and will complicate the user experience of
+contributing to Catalogs.
+
+### Releases - Repositories; No Catalogs
+
+We could require that each resource is defined in its own repository. This will ensure that each resource will
+have its own tag, hence its own versioning. Each resource will be shared directly in the Hub, without needing a
+centralized Catalog. However, this will be a drastic change to the Catalog organization and will complicate the
+user experience of contributing resources.
```

Comment:
@wlynch. There are 2 points that I can think of:
1. Comparing to the monthly release tagging approach described above, we could not enforce contributors to tag their resources (at least in a consistent way) in their own repo, so it is hard to maintain a consistent mapping from hub version to catalog version (e.g. tag "v0.7" in a Catalog == version "0.7" in the Hub)

2. Currently, Tekton hub uses a cron job to fetch resources from the Catalog repo. If we deprecate the catalog repo, the contributor needs to "register" their own repo in the Hub (by creating an extra PR in the Hub repo) so that the new resource can be discovered. This may increase end-to-end latency for a new resource showing up in the UI and I think it also complicates the UX as well?

One more concern I have with separated repo is that: once a new repo is registered in the hub, the contributor can do whatever changes he/she wants to do in their own repo and we don't have control to it at all (delete the repo, publish any content...)? I feel it also introduces the maintenance cost and security risk in the Hub.

Reply:
I am big +1 in favor of splitting the catalog organization to multiple repos, this is going to be a huge change for hub and contribution experience but much more sustainable in the future.

There is a prior example doing this with flathub, https://github.com/flathub/. There is still a app submission process that needs to be validated but let the user update their own: https://github.com/flathub/flathub/wiki/App-Submission 

I believe CI/CD can take care of creating a global manifest file to centralize the new tasks or version updated for UIs (ie: hub/cli) to consume.

> One more concern I have with separated repo is that: once a new repo is registered in the hub, the contributor can do whatever changes he/she wants to do in their own repo and we don't have control to it at all (delete the repo, publish any content...)? I feel it also introduces the maintenance cost and security risk in the Hub.

that security risk has always been there, as a catalog task reviewer, we don't have the capability to actually try the tasks, only can review the task structure conformance  and maybe general tekton tasks advice on task. But the image may be able to contains anything. Aside of maybe the "git-clone" or tasks using tektoncd/pipelines container images i don't think there is any tasks in tektoncd/catalog that can be properly vetted as "secure" (the same way you can't really trust any maven/node/pypi/go.mod deps).

Reply:
@wlynch the UX comment was mostly driven by not knowing how contributors can run tests for their resources - this is run by tekton today - so will contributors use their own infra? and handle their own testing? (guess it'd also be up to them to handle linting - which was supported through catlin)

Reply:
another holdback is how we migrate the existing tektoncd/catalog to separate repositories - if this is the path we take, then the official resources are the only ones that'd like tektoncd github org - so where do the other ones move to? do we create a separate github org for them as a transition plan for resource we already have (but new resources are in other orgs)?

Reply:
@QuanZhang-William:

> Comparing to the monthly release tagging approach described above, we could not enforce contributors to tag their resources (at least in a consistent way) in their own repo, so it is hard to maintain a consistent mapping from hub version to catalog version (e.g. tag "v0.7" in a Catalog == version "0.7" in the Hub)

For ensuring consistent versioning semantics so that the hub can process things consistently, could we document/enforce "you must use semver or else the hub will not recognize the version"? I'm imagining something similar to https://go.dev/doc/modules/version-numbers.

> Currently, Tekton hub uses a cron job to fetch resources from the Catalog repo. If we deprecate the catalog repo, the contributor needs to "register" their own repo in the Hub (by creating an extra PR in the Hub repo) so that the new resource can be discovered. This may increase end-to-end latency for a new resource showing up in the UI and I think it also complicates the UX as well?

That's probably okay in most cases - IIUC the registration step would only need to happen once, then we could use the same cron approach as before?

Something that might be a good source of inspiration is [`pkg.go.dev`](https://pkg.go.dev/about), since I think there are quite a few parallels here for versioning, organization, discoverability, etc. I think they do something nifty where if a module is requested and it isn't in `pkg.go.dev` yet, it enqueues to be fetched.

There's probably more value if we could leverage the catalog/hub with something like the [Go module proxy](https://proxy.golang.org/). The mirroring + checksumming could help avoid some of the deletion / force push concerns, and it would also be a good opportunity to get metrics on public Task usage.

---
 
@jerop: 

> so will contributors use their own infra? and handle their own testing? 

Yes! This is probably a good thing - if Tekton maintainers aren't reviewing/maintaining community contributions, then separating out and isolating community repo CI is likely better for reducing risk to Tekton infra and other community Tasks.

> (guess it'd also be up to them to handle linting - which was supported through catlin)

Maybe? Few ways you can approach this depending on how strict of a requirement you want to enforce:

- Premade Tasks for running linting (make it easy for people to run linting, but don't make it a strict requirement)
- Run a GitHub App that will run linting for projects in a Tekton community controlled environment

> so where do the other ones move to? do we create a separate github org for them as a transition plan for resource we already have (but new resources are in other orgs)?

Personally I'd say open up GitHub issues for each Task and tag the current owners. Give people time to migrate Tasks to their own repos (+ maybe give them a script to make migration easy), then after a certain period of time archive the old repo.

Reply:
thanks for the discussion here @wlynch @QuanZhang-William @chmouel -- updated the TEP to reflect the points raised here, please take a look 🙏🏾

Reply:
thanks for the discussion here @wlynch @QuanZhang-William @chmouel -- updated the TEP to reflect the points raised here, please take a look 🙏🏾